Feat/docs and precommit (#4)
ukhsa-collaboration/mscape-template: README.md

# mscape-template

This repository is a template for creating new repositories containing code that will run on mSCAPE. It serves as a
guide for code layout and files will need amending to fit the repo purpose.

## How to use this template:
If you are reading this README...
* **whilst in the original `mscape-template` repository**, then you can use this template to
start a new mSCAPE repo. For example, if you are on the repo page on GitHub in the browser right now, you can click the
green <span style="color:green">use this template</span> button on the top right. If you want to use this template in a
new local repo, just copy the contents of this repo into your new repo.

* **inside your own repo**, before committing to this repo, you **MUST** do 2 things - set up branch protection on the
remote and install the predefined pre-commit hook.

1.) Set up branch protection:

    a. You should have either git cloned this repo locally, but if not, you will need the file
    `recommended-branch-protection.json` from the template repo.

    b. Go to settings in the GitHub repo in your browser. The Settings tab is under the repo name and search bar.

>     c. On the banner on the right, you should see 'Rules'. Clickon the drop down, and click 'Rulesets'.

Reviewer: ```suggestion
    c. On the banner on the right, you should see 'Rules'. Click on the drop down, and click 'Rulesets'.
```

    d. Click on the green 'New Ruleset' button in the top right, and select 'Import a ruleset'.

    e. A file browser window will open, where you can select the `recommended-branch-protection.json` from your locally

    cloned repo or from the file you downloaded. Click 'open' in the file browser window to upload it.
    f. Scroll down and click the green 'Create' button.

>     g. when clicking back into the 'rulesets' area from the banner on the right, you should now see

Reviewer: ```suggestion
    g. When clicking back into the 'rulesets' area from the banner on the right, you should now see
```

It's a good idea to click through the rules implemented to understand what is happening.
Essentially, when applied to the `main` branch:
- Restrict deletions - can't delete `main` (only those with bypass can)
- Require a pull request before merging - Require all commits be made to a non-target branch and submitted via a pull
request before they can be merged.
- 1 required approval before pull request can be merged.
- Dismiss stale pull request approvals when new commits are pushed - New, reviewable commits pushed will dismiss
previous pull request review approvals.
- Require approval of the most recent reviewable push - most recent reviewable push must be approved by someone other
than the person who pushed it.
- Require status checks to pass - must pass tests and linting as laid out in the `.github/workflows/test-build.yml`.
- Require branches to be up to date before merging - pull requests targeting `main` must be tested with the latest code.
- Block for pushes - prevent users with push access from force pushing to `main`.

2.) The pre-commit hook. This will run ruff, some Python linting and a search for potential climb-IDs* every time you
commit locally.

>     a. First [install pre-commit](https://pre-commit.com/#install):

Reviewer: Hyperlink formatting doesn't seem to be working correctly in md view

        pip install pre-commit

>     b. install the pre-commit hook into your .git/hooks directory using the pre-defined config in

Reviewer: ```suggestion
    b. Install the pre-commit hook into your .git/hooks directory using the pre-defined config in
```

    .pre-commit-config.yaml:
        pre-commit install

>     c. (optional) run pre-commit against all files, to see if the current files pass the hook:

Reviewer: ```suggestion
    c. (optional) Run pre-commit against all files, to see if the current files pass the hook:
```

        pre-commit run --all-files

>     Now, whenever you run git commit the pre-commit hook should now run automatically.

Reviewer: ```suggestion
    Now, whenever you run git commit the pre-commit hook should run automatically.
```

> _\* Note on the climb-id search - this will match any occurence of C- followed by 5 characters of A-F or 0-9. This

Reviewer: ```suggestion
_\* Note on the climb-id search - this will match any occurrence of C- followed by 5 characters of A-F or 0-9. This
```

> __may__ match fasta sequences, such as 'AAC-CCAC...'. In this case, you may need to commit this seperately and remove

Reviewer: ```suggestion
__may__ match fasta sequences, such as 'AAC-CCAC...'. In this case, you may need to commit this separately and remove
```

> __may__ match fasta sequences, such as 'AAC-CCAC...'. In this case, you may need to commit this seperately and remove

Reviewer: May also need to account to test files with fake climb ids/changing the format of these to note be caught by the hook?

# mscape-template

This repository is a template for creating new repositories containing code that will run on mSCAPE. It serves as a
guide for code layout and files will need amending to fit the repo purpose.

## How to use this template:
If you are reading this README...
* **whilst in the original `mscape-template` repository**, then you can use this template to
start a new mSCAPE repo. For example, if you are on the repo page on GitHub in the browser right now, you can click the
green <span style="color:green">use this template</span> button on the top right. If you want to use this template in a
new local repo, just copy the contents of this repo into your new repo.

* **inside your own repo**, before committing to this repo, you **MUST** do 2 things - set up branch protection on the
remote and install the predefined pre-commit hook.

1.) Set up branch protection:

a. You should have either git cloned this repo locally, but if not, you will need the file

>     `recommended-branch-protection.json` from the template repo.

Reviewer: Fix formatting?

b. Go to settings in the GitHub repo in your browser. The Settings tab is under the repo name and search bar.

c. On the banner on the right, you should see 'Rules'. Click on the drop down, and click 'Rulesets'.

d. Click on the green 'New Ruleset' button in the top right, and select 'Import a ruleset'.

e. A file browser window will open, where you can select the `recommended-branch-protection.json` from your locally 
cloned repo or from the file you downloaded. Click 'open' in the file browser window to upload it.

f. Scroll down and click the green 'Create' button.

g. When clicking back into the 'rulesets' area from the banner on the right, you should now see `protect-main` with 4
rules, targetting 1 branch (which should be main).

> It's a good idea to click through the rules implemented to understand what is happening.

Reviewer: Think this and below sentence can be condensed down - something like 'Once set up, the branch protections enforce the following:'

Essentially, when applied to the `main` branch:
- Restrict deletions - can't delete `main` (only those with bypass can)

> - Require a pull request before merging - Require all commits be made to a non-target branch and submitted via a pull

Reviewer: Should this be two separate points?

request before they can be merged.
- 1 required approval before pull request can be merged.
- Dismiss stale pull request approvals when new commits are pushed - New, reviewable commits pushed will dismiss
previous pull request review approvals.
- Require approval of the most recent reviewable push - most recent reviewable push must be approved by someone other
than the person who pushed it.
- Require status checks to pass - must pass tests and linting as laid out in the `.github/workflows/test-build.yml`.
- Require branches to be up to date before merging - pull requests targeting `main` must be tested with the latest
code.
- Block for pushes - prevent users with push access from force pushing to `main`.

> 2.) The pre-commit hook. This will run ruff, some Python linting and a search for potential climb-IDs* every time you

Reviewer: ```suggestion
2.) Set up pre-commit hooks. This will run ruff, some Python linting and a search for potential climb-IDs* every time you
```

commit locally.

a. First [install pre-commit](https://pre-commit.com/#install):
    
    pip install pre-commit

b. Install the pre-commit hook into your .git/hooks directory using the pre-defined config in .pre-commit-config.yaml:
        
    pre-commit install

> c. (optional) Run pre-commit against all files, to see if the current files pass the hook:

Reviewer: I think maybe remove the optional?

- Code to be in src/ layout
- `tests/` folder at same level as `src/`
- .github/ folder containing workflows and a pull request template
- .gitignore file
- pyproject.toml file
- .pre-commit-hooks.yaml

This repo follows the above structure and contains examples of the file referenced above.

### Commit practices
Commits **MUST** be:

* small and focused
* clearly described

You **may** wish to use the [conventional commits format](https://www.conventionalcommits.org/en/v1.0.0/).

### Bramching Strategy
UKHSA Engineering Standards suggest we **should** follow the
[GitHub Flow-style](https://docs.github.com/en/get-started/using-github/github-flow) branching model.

    feature/ — functional changes
    bugfix/ — defect fixes
    spike/ — experimental work (not merged)
    hotfix/ — urgent fixes to production

### Branch protection
Branch protection should be inplace, preventing pushing to default branch (usually main), requiring reviewed pull 
requests etc.

[Read the UKHSA Engineering Standards guidelines for more information.](https://ukhsa-collaboration.github.io/standards-org/development-standards/appendix/branch-protection-rules/)

### Pull Requests and merge process

> When you have a feature or fix ready, you can submit a merge request.

Reviewer: ```suggestion
When you have a feature or fix ready, you can submit a pull request.
```

> The merge request should be populated using the merge-request template. This include a checklist, and can be 

Reviewer: ```suggestion
The pull request should be populated using the pull request template. This includes a checklist, and can be 
```
